File: server/src/functions/epostAddress.ts
import { fail, ok, type FnHandler } from './types.ts'

const ENDPOINT =
  'http://openapi.epost.go.kr/postal/retrieveNewAdressAreaCdSearchAllService/retrieveNewAdressAreaCdSearchAllService/getNewAddressListAreaCdSearchAll'

const ITEM_RE = /<newAddressListAreaCdSearchAll>([\s\S]*?)<\/newAddressListAreaCdSearchAll>/g
const HEADER_RE = /<cmmMsgHeader>([\s\S]*?)<\/cmmMsgHeader>/
const FIELD_RE = {
  successYN: /<successYN>([\s\S]*?)<\/successYN>/,
  errMsg: /<errMsg>([\s\S]*?)<\/errMsg>/,
  returnCode: /<returnCode>([\s\S]*?)<\/returnCode>/,
  zipNo: /<zipNo>([\s\S]*?)<\/zipNo>/,
  lnmAdres: /<lnmAdres>([\s\S]*?)<\/lnmAdres>/,
  rnAdres: /<rnAdres>([\s\S]*?)<\/rnAdres>/,
}

function serviceKey() {
  const raw = (process.env.EPOST_SERVICE_KEY ?? '').trim()
  if (!raw) throw new Error('우체국 Open API 키(EPOST_SERVICE_KEY)가 설정되지 않았습니다.')
  return raw.includes('%') ? decodeURIComponent(raw) : raw
}

function unwrap(value: string) {
  return value
    .replace(/<!\[CDATA\[([\s\S]*?)\]\]>/g, '$1')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&amp;/g, '&')
    .trim()
}

function field(chunk: string, re: RegExp) {
  const match = re.exec(chunk)
  return match ? unwrap(match[1]) : ''
}

function parseResults(xmlText: string) {
  const header = HEADER_RE.exec(xmlText)?.[1]
  if (header) {
    const success = field(header, FIELD_RE.successYN)
    if (success === 'N') {
      const err = field(header, FIELD_RE.errMsg)
      const code = field(header, FIELD_RE.returnCode)
      throw new Error(err || `우체국 주소 조회 실패 (코드 ${code || '?'})`)
    }
  }

  const results = []
  for (const match of xmlText.matchAll(ITEM_RE)) {
    const chunk = match[1]
    const zipNo = field(chunk, FIELD_RE.zipNo)
    const roadAddress = field(chunk, FIELD_RE.lnmAdres)
    const jibunAddress = field(chunk, FIELD_RE.rnAdres)
    if (!zipNo && !roadAddress && !jibunAddress) continue
    results.push({
      id: `epost:${zipNo}:${roadAddress}:${jibunAddress}`,
      zipNo,
      roadAddress,
      jibunAddress,
    })
  }
  return results
}

async function search(query: string, countPerPage: number, currentPage: number) {
  const url = new URL(ENDPOINT)
  url.searchParams.set('serviceKey', serviceKey())
  url.searchParams.set('srchwrd', query)
  url.searchParams.set('countPerPage', String(Math.max(1, Math.min(countPerPage, 50))))
  url.searchParams.set('currentPage', String(Math.max(1, currentPage)))

  const response = await fetch(url)
  if (!response.ok) throw new Error(`우체국 주소 조회 HTTP ${response.status}`)
  return parseResults(await response.text())
}

export const epostAddress: FnHandler = async (ctx) => {
  if (!ctx.userId) return fail('로그인이 필요합니다.', 401)
  try {
    const query = String(ctx.body?.query ?? '').trim()
    if (query.length < 2) return fail('검색어를 입력해 주세요.')
    const countPerPage = Number(ctx.body?.countPerPage ?? 20)
    const currentPage = Number(ctx.body?.currentPage ?? 1)
    if (!Number.isFinite(countPerPage) || !Number.isFinite(currentPage)) {
      return fail('요청이 올바르지 않습니다.')
    }
    return ok({
      results: await search(query, countPerPage, currentPage),
    })
  } catch (error) {
    return fail(error instanceof Error ? error.message : '우체국 주소 조회에 실패했습니다.', 500)
  }
}

File: server-py/app/functions/epost_address.py
"""
우체국(우정사업본부) 통합검색 5자리 우편번호 조회.

공공데이터포털 '우편번호 정보조회' Open API 프록시.
브라우저에 serviceKey 를 노출하지 않기 위해 서버에서만 호출한다.
"""
from __future__ import annotations

import os
import re
import xml.etree.ElementTree as ET
from urllib.parse import unquote

import httpx

from .types import FnCtx, FnResult, fail, ok

ENDPOINT = (
    "http://openapi.epost.go.kr/postal/"
    "retrieveNewAdressAreaCdSearchAllService/"
    "retrieveNewAdressAreaCdSearchAllService/"
    "getNewAddressListAreaCdSearchAll"
)

_TAG = re.compile(r"\{.*\}")


def _local(tag: str) -> str:
    return _TAG.sub("", tag)


def _text(node: ET.Element | None) -> str:
    if node is None or node.text is None:
        return ""
    return node.text.strip()


def _service_key() -> str:
    raw = (os.environ.get("EPOST_SERVICE_KEY") or "").strip()
    if not raw:
        raise Exception("우체국 Open API 키(EPOST_SERVICE_KEY)가 설정되지 않았습니다.")
    # 포털에서 복사한 인코딩 키(%2B 등)를 한 번 풀어 httpx 가 다시 인코딩하게 한다.
    return unquote(raw) if "%" in raw else raw


def _parse_results(xml_text: str) -> list[dict]:
    root = ET.fromstring(xml_text)
    header = None
    items: list[ET.Element] = []
    for child in root:
        name = _local(child.tag)
        if name == "cmmMsgHeader":
            header = child
        elif name == "newAddressListAreaCdSearchAll":
            items.append(child)

    if header is not None:
        success = _text(next((c for c in header if _local(c.tag) == "successYN"), None))
        if success == "N":
            err = _text(next((c for c in header if _local(c.tag) == "errMsg"), None))
            code = _text(next((c for c in header if _local(c.tag) == "returnCode"), None))
            raise Exception(err or f"우체국 주소 조회 실패 (코드 {code or '?'})")

    results: list[dict] = []
    for item in items:
        fields = {_local(c.tag): _text(c) for c in item}
        zip_no = fields.get("zipNo", "")
        # 이 서비스 명세: lnmAdres=도로명, rnAdres=지번
        road = fields.get("lnmAdres", "")
        jibun = fields.get("rnAdres", "")
        if not zip_no and not road and not jibun:
            continue
        results.append(
            {
                "id": f"epost:{zip_no}:{road}:{jibun}",
                "zipNo": zip_no,
                "roadAddress": road,
                "jibunAddress": jibun,
            }
        )
    return results


async def _search(query: str, count_per_page: int, current_page: int) -> list[dict]:
    params = {
        "serviceKey": _service_key(),
        "srchwrd": query,
        "countPerPage": str(max(1, min(count_per_page, 50))),
        "currentPage": str(max(1, current_page)),
    }
    async with httpx.AsyncClient(timeout=20.0) as client:
        response = await client.get(ENDPOINT, params=params)
        response.raise_for_status()
        return _parse_results(response.text)


async def epost_address(ctx: FnCtx) -> FnResult:
    if not ctx.user_id:
        return fail("로그인이 필요합니다.", 401)
    try:
        query = (ctx.body.get("query") or "").strip()
        if len(query) < 2:
            return fail("검색어를 입력해 주세요.")
        count = int(ctx.body.get("countPerPage") or 20)
        page = int(ctx.body.get("currentPage") or 1)
        return ok({"results": await _search(query, count, page)})
    except ValueError:
        return fail("요청이 올바르지 않습니다.")
    except Exception as error:  # noqa: BLE001
        return fail(str(error) or "우체국 주소 조회에 실패했습니다.", 500)
